test/sync/src/androidTest/kotlin/io/realm/test/mongodb/shared/UserTests.kt

```diff
@@ -0,0 +1,449 @@
+        if (this::app.isInitialized) {
+            app.asTestApp.close()
+        }
+    }
+
+    @Test
+    fun getApp() = runBlocking {
+        val anonUser = app.login(Credentials.anonymous())
+        assertEquals(anonUser.app, app.asTestApp.app)
+    }
+
+    @Test
+    fun getState_anonymousUser() = runBlocking {
+        val anonUser = app.login(Credentials.anonymous())
+        assertEquals(User.State.LOGGED_IN, anonUser.state)
+        anonUser.logOut()
+        assertEquals(User.State.REMOVED, anonUser.state)
+    }
+
+    @Test
+    fun getState_emailUser() = runBlocking {
+        val emailUser = createUserAndLogin()
+        assertEquals(User.State.LOGGED_IN, emailUser.state)
+        emailUser.logOut()
+        assertEquals(User.State.LOGGED_OUT, emailUser.state)
+        // TODO wait for EmailPasswordAuth
+//        app.removeUser(emailUser)
+//        assertEquals(User.State.REMOVED, emailUser.state)
+    }
+
+    @Test
+    fun logOut() = runBlocking {
+        val anonUser = app.login(Credentials.anonymous())
+
+        // Anonymous users are removed upon log out
+        assertEquals(anonUser, app.currentUser)
+        assertEquals(User.State.LOGGED_IN, anonUser.state)
+        anonUser.logOut()
+        assertEquals(User.State.REMOVED, anonUser.state)
+        assertNull(app.currentUser)
+
+        // Users registered with Email/Password will register as Logged Out
+        val user2 = createUserAndLogin()
+        val current: User = app.currentUser!!
+        assertEquals(user2, current)
+        user2.logOut()
+        assertEquals(User.State.LOGGED_OUT, user2.state)
+        // Same effect on all instances
+        assertEquals(User.State.LOGGED_OUT, current.state)
+        // And no current user anymore
+        assertNull(app.currentUser)
+    }
+
+    @Test
+    fun logOutUserInstanceImpactsCurrentUser() = runBlocking {
+        val anonUser = app.login(Credentials.anonymous())
+
+        val currentUser = app.currentUser!!
+        assertEquals(User.State.LOGGED_IN, currentUser.state)
+        assertEquals(User.State.LOGGED_IN, anonUser.state)
+        assertEquals(currentUser, anonUser)
+
+        anonUser.logOut()
+
+        assertNotEquals(User.State.LOGGED_OUT, currentUser.state)
+        assertNotEquals(User.State.LOGGED_OUT, anonUser.state)
+        assertNull(app.currentUser)
+    }
+
+    @Test
+    fun logOutCurrentUserImpactsOtherInstances() = runBlocking {
+        val anonUser = app.login(Credentials.anonymous())
+
+        val currentUser = app.currentUser!!
+        assertEquals(User.State.LOGGED_IN, currentUser.state)
+        assertEquals(User.State.LOGGED_IN, anonUser.state)
+        assertEquals(currentUser, anonUser)
+
+        currentUser.logOut()
+
+        assertNotEquals(User.State.LOGGED_OUT, currentUser.state)
+        assertNotEquals(User.State.LOGGED_OUT, anonUser.state)
+        assertNull(app.currentUser)
+    }
+
+//    @Test
+//    fun repeatedLogInAndOut() = runBlocking {
```

Comment:
I guess we have all APIs to include this

Reply:
Not really, we are missing `User.profile` but I think it's still valid to enable it using the email and password used to create the user.

Reply:
Sorry, missed that. I guess whatever makes most sense for you then.